fs: nvs: Add recovery path for corrupted NVS

corymayer opened the pull request:
When all sectors of the settings partition have some data that is not 0xFF, NVS will fail to initialize settings. Add a recovery path to erase the whole settings partition and recover. Should not happen unless there's some serious flash corruption, or possibly interrupted garbage collection. Verified by recreating the scenario with flash CLIs.

corymayer commented:
> Hi @corymayer, this is a nice addition. Thank you for the proposal.
> 
> If you agree I would like to see this possibility enabled by a Kconfig option.
> 
> I don't think the change in comment is required, when this problem occurs it is not by definition a corrupt NVS system, it can also be a non NVS partition.

Sure, will update soon

Laczen commented on subsys/fs/nvs/Kconfig:
```diff
@@ -37,6 +37,14 @@ config NVS_DATA_CRC
 	  The CRC-32 is transparently stored at the end of the data field,
 	  in the NVS data section, so 4 more bytes are needed per NVS element.
 
+config NVS_BAD_PARTITION_RECOVERY
```
What do you think about "NVS_INIT_BAD_REGION" ?

I would like this option and help text to reflect what it does and not (less) why it does this. NVS has no knowledge of partitions so using "partition" should be avoided.

The help could be something like:
Enable automatic initialization of a NVS on a flash region that does not contain a valid NVS. A region containing an invalid NVS can be caused by corruption or by providing a non-empty region. This option ensures a new NVS can be created.

corymayer replied:
Makes sense, updated

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -781,10 +781,21 @@ static int nvs_startup(struct nvs_fs *fs)
+#ifdef CONFIG_NVS_BAD_PARTITION_RECOVERY
+		LOG_WRN("All sectors closed, erasing all settings...");
+		addr = 0U;
+		rc = flash_erase(fs->flash_device, 0, fs->sector_size * fs->sector_count);
```
I don't think this is erasing at the correct location.

rghaddab replied:
You should use flash_flatten instead.

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -781,10 +781,21 @@ static int nvs_startup(struct nvs_fs *fs)
+		if (rc) {
+			rc = -EIO;
+			goto end;
+		}
```
please insert a space after this line.

corymayer replied:
Line was removed

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -781,10 +781,21 @@ static int nvs_startup(struct nvs_fs *fs)
+			rc = -EIO;
+			goto end;
+		}
+		i = 0;
```
I think we want:
`i = fs->sector_count;` 
`addr = ((fs->sector_count - 1) << ADDR_SECT_SHIFT) + (uint16_t)(fs->sector_size - ate_size);`

rghaddab commented on subsys/fs/nvs/nvs.c:
```diff
@@ -781,10 +781,21 @@ static int nvs_startup(struct nvs_fs *fs)
+	/* all sectors are closed, this is not a nvs fs or irreparably corrupted */
 	if (closed_sectors == fs->sector_count) {
+#ifdef CONFIG_NVS_BAD_PARTITION_RECOVERY
+		LOG_WRN("All sectors closed, erasing all settings...");
```
There should be no mention of settings here.
NVS and Settings are two different subsystems.
You should use "erase all entries" or maybe "erasing all sectors"

rghaddab commented on subsys/fs/nvs/nvs.c:
```diff
@@ -781,10 +781,21 @@ static int nvs_startup(struct nvs_fs *fs)
+		addr = 0U;
+		rc = flash_erase(fs->flash_device, 0, fs->sector_size * fs->sector_count);
+		if (rc) {
+			rc = -EIO;
```
why do you need to change the error code returned by the above function. Just keep the same error code

Damian-Nordic commented on subsys/fs/nvs/nvs.c:
```diff
@@ -781,10 +781,20 @@ static int nvs_startup(struct nvs_fs *fs)
 	if (closed_sectors == fs->sector_count) {
+#ifdef CONFIG_NVS_INIT_BAD_REGION
+		LOG_WRN("All sectors closed, erasing all sectors...");
+		addr = 0U;
```
Was this code tested? If you enter this `if` and you erase the sectors, then you also enter the `if (i == fs->sector_count)` condition below, where there is `rc = nvs_flash_cmp_const(fs, addr - ate_size, erase_value, sizeof(struct nvs_ate));` that assumes that `addr` points to the last ATE within the last sector. Am I correct?

Please consider extending NVS unit tests with a scenario that you initialize NVS using a garbage data in all sectors.

corymayer replied:
Just did a full re-test with the latest version I pushed. Set the ATEs to 0xdeadbeef and verified the log is printed and the NVS settings are written correctly afterwards.

Can you point me to the unit tests for NVS? I don't see them in this directory.

Damian-Nordic replied:
Thank you. The tests are in `tests/subsys/fs/nvs` directory. Though feel free to merge this PR and add the tests later :)

corymayer commented on subsys/fs/nvs/nvs.c:
```diff
@@ -781,10 +781,20 @@ static int nvs_startup(struct nvs_fs *fs)
 	if (closed_sectors == fs->sector_count) {
+#ifdef CONFIG_NVS_INIT_BAD_REGION
+		LOG_WRN("All sectors closed, erasing all sectors...");
+		addr = 0U;
```
Missed this, we can just remove this line too right, since the previous loop will ensure addr is set to the last sector?

Laczen replied:
I think it is set to the first sector when all sectors are closed.

corymayer replied:
I see, updated to set them to the last sector as you suggested

Laczen reviewed (dismissed):
@corymayer thank you for this PR.

rghaddab commented on subsys/fs/nvs/Kconfig:
```diff
@@ -37,6 +37,14 @@ config NVS_DATA_CRC
+	  Enable automatic initialization of a NVS on a flash region that does
+	  not contain a valid NVS. A region containing an invalid NVS can be
+	  caused by corruption or by providing a non-empty region. This option
+	  ensures a new NVS can be created.
```
The naming here is confusing. You are talking about region but in the description you are referring to "bad partition recovery"
AFAIK the term "Region" doesn't seem to be used anywhere in NVS.
NVS knows only about Sectors of data
Could be NVS_INIT_BAD_SECTORS

Laczen replied:
The terminology `region` was carefully chosen not to overlap with partition (as defined in dts), area (as used in `flash_area_...` routines, there also are no bad sectors (which would be used by nand devices).

@corymayer, please update the bool to "Non-volatile Storage bad region recovery".

Laczen replied:
"Memory region" is also existing terminology.

rghaddab replied:
"memory region" is used to describe  a part of the memory. This is what this PR is about, right ? Am I missing something ? Or do you mean that the "memory region" terminology is reserved for other purposes in zephyr ?

de-nordic replied:
Everything is existing terminology, so context matters. If something does not exist in a given context than it can be used to describe some characteristic in that context. And NAND actually uses Bad Block Management, for example winbond devices, to manage sectors in pages, which is again vendor specific nomenclature, because sectors/pages/block naming varies between devices.

So, Region or Memory region, when  defined within context of NVS as a specific offset/range tuple may be fine, as long as it's relation to sectors or partitions, as used by NVS, is clear.

Laczen replied:
"memory region" is used in dts for linker memory regions, in user mode for memory domains... It is in no way reserved exclusively for these purposes but might be confusing to use it here as some people associate memory with RAM.

rghaddab replied:
In that case I prefer "memory region" which covers (flash/ram/... regions) instead of just "region"
but that's not a blocker for me.

corymayer replied:
Updated the name to "bad memory region" and added a unit test

Laczen commented:
@corymayer are you a discord user? I would like to get in touch.

corymayer commented:
> @corymayer are you a discord user? I would like to get in touch.

Yes, just joined as corymeta_74139

Laczen reviewed (dismissed):
I find the term "bad memory region" confusing and long to set as config variable, but I'm not going to spend more time on terminology.

Thanks for adding the test.

corymayer commented:
Fixed an off-by-one in the unit test

Laczen commented on tests/subsys/fs/nvs/src/main.c:
```diff
@@ -959,3 +959,40 @@ ZTEST_F(nvs, test_nvs_cache_hash_quality)
+		err = flash_write(fixture->fs.flash_device, fixture->fs.offset +
+						 (fixture->fs.sector_size * (i + 1)) -
+						  sizeof(struct nvs_ate), &data, sizeof(data));
+		zassert_true(err == 0,  "flash_write failed: %d", err);
```
One space to much after the ",".

corymayer replied:
Fixed these

Laczen commented on tests/subsys/fs/nvs/src/main.c:
```diff
@@ -959,3 +959,40 @@ ZTEST_F(nvs, test_nvs_cache_hash_quality)
+
+#ifdef CONFIG_NVS_INIT_BAD_MEMORY_REGION
+	err = nvs_mount(&fixture->fs);
+	zassert_true(err == 0,  "nvs_mount call failure: %d", err);
```
one space to much after ",".

Laczen commented on tests/subsys/fs/nvs/src/main.c:
```diff
@@ -959,3 +959,40 @@ ZTEST_F(nvs, test_nvs_cache_hash_quality)
+	execute_long_pattern_write(TEST_DATA_ID, &fixture->fs);
+#else
+	err = nvs_mount(&fixture->fs);
+	zassert_true(err == -EDEADLK,  "nvs_mount call ok, expect fail: %d", err);
```
One space to much after ",".

Laczen commented on tests/subsys/fs/nvs/src/main.c:
```diff
@@ -959,3 +959,40 @@ ZTEST_F(nvs, test_nvs_cache_hash_quality)
+ */
+ZTEST_F(nvs, test_nvs_init_bad_memory_region)
+{
+
```
NIT: remove this line

Laczen commented on tests/subsys/fs/nvs/src/main.c:
```diff
@@ -959,3 +959,40 @@ ZTEST_F(nvs, test_nvs_cache_hash_quality)
+	/* Write bad ATE to each sector */
+	for (uint16_t i = 0; i < TEST_SECTOR_COUNT; i++) {
+		data = 0xdeadbeef;
+		err = flash_write(fixture->fs.flash_device, fixture->fs.offset +
```
NIT: improve the alignment or introduce a `offset` variable to make this line shorter.

Laczen commented on tests/subsys/fs/nvs/src/main.c:
```diff
@@ -959,3 +959,40 @@ ZTEST_F(nvs, test_nvs_cache_hash_quality)
+	for (uint16_t i = 0; i < TEST_SECTOR_COUNT; i++) {
+		data = 0xdeadbeef;
+		err = flash_write(fixture->fs.flash_device,
+			fixture->fs.offset +
```
The alignment should be with the start of `fixture->fs.flash_device`. The complete file however is no example for correct alignment :-).

corymayer committed:
fs: nvs: Add recovery path for corrupted NVS

When all sectors have non-zero data, NVS would fail to init.
Add recovery path option to erase the all sectors and reinitialize.
This could occur due to non-empty sectors, or corrupted data.

Signed-off-by: Cory Andrew Mayer <corymayer@meta.com>

corymayer committed:
tests: subsys/fs/nvs: add test for initializing bad region

Add unit test to exercise bad memory region recovery path

Signed-off-by: Cory Andrew Mayer <corymayer@meta.com>

corymayer commented:
> @corymayer Please fix the compliance.

I was able to run the compliance check locally and verify all issues are resolved

Laczen commented on tests/subsys/fs/nvs/src/main.c:
```diff
@@ -959,3 +959,40 @@ ZTEST_F(nvs, test_nvs_cache_hash_quality)
+		data = 0xdeadbeef;
+		err = flash_write(fixture->fs.flash_device,
+				  fixture->fs.offset + (fixture->fs.sector_size * (i + 1)) -
+					  sizeof(struct nvs_ate),
```
NIT, I think this alignment might still be considered incorrect.
